base_layer/core/src/proof_of_work/proof_of_work.rs

```diff
@@ -0,0 +1,318 @@
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    blocks::BlockHeader,
+    proof_of_work::{blake_pow::blake_difficulty, monero::monero_difficulty, Difficulty},
+};
+use bytes::{self, BufMut};
+use serde::{Deserialize, Serialize};
+use std::{
+    convert::TryFrom,
+    fmt::{Display, Error, Formatter},
+};
+use tari_utilities::hex::Hex;
+
+pub trait AchievedDifficulty {}
+
+#[repr(u8)]
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+pub enum PowAlgorithm {
+    Monero = 0,
+    Blake = 1,
+}
+
+/// Used to compare proof of work difficulties without scaling factors
+#[derive(Debug, Clone, PartialEq)]
+pub enum Ordering {
+    GreaterThan,
+    LessThan,
+    Equal,
+    Indeterminate,
+}
+
+impl TryFrom<u64> for PowAlgorithm {
+    type Error = String;
+
+    fn try_from(v: u64) -> Result<Self, Self::Error> {
+        match v {
+            0 => Ok(PowAlgorithm::Monero),
+            1 => Ok(PowAlgorithm::Blake),
+            _ => Err("Invalid PoWAlgorithm".into()),
+        }
+    }
+}
+
+/// The proof of work data structure that is included in the block header. There's some non-Rustlike redundancy here
+/// to make serialization more straightforward
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ProofOfWork {
+    /// The total accumulated difficulty for each proof of work algorithms for all blocks since Genesis,
+    /// but not including this block, tracked separately.
+    pub accumulated_monero_difficulty: Difficulty,
+    pub accumulated_blake_difficulty: Difficulty,
```

Comment:
should we not rather move this over to an array of difficulties. If our simulations point to more than 2 it might be redone.

Reply:
I've considered this; I started out with an array, and it's much cleaner this way after I refactored.
As long as we stay under 5 PoW's I'm confident this approach is more readable . Five algos might be an inflection point, but I think we've decided that 3 is the highest we'd consider for the forseeeable future.